llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,128 @@
+                             "required for funtions with external linkage"),
+                    cl::init(""));
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+      continue;
+
+    if (F.hasFnAttribute("sycl-call-if-on-device-conditionally") &&
+        ("true" == F.getFnAttribute("sycl-call-if-on-device-conditionally")
+                       .getValueAsString()))
+      FCallers.push_back(&F);
+  }
+
+  // A vector instead of DenseMap to make LIT tests predictable
+  SmallVector<std::pair<Function *, Function *>, 8> FCallersToFActions;
+  for (Function *FCaller : FCallers) {
+    // Find call to @CallableXXX in call_if_on_device_conditionally function
+    // (FAction). FAction should be a literal (i.e. not a pointer). The
+    // structure of the header file ensures that there is exactly one such
+    // instruction.
+    for (Instruction &I : instructions(FCaller)) {
+      if (auto *CI = dyn_cast<CallInst>(&I);
+          CI && (Intrinsic::IndependentIntrinsics::not_intrinsic ==
+                 CI->getIntrinsicID())) {
+        FCallersToFActions.push_back(
+            std::make_pair(FCaller, CI->getCalledFunction()));
+        break;
+      }
+    }
+  }
+
+  int FCallerIndex = 1;
+  for (const auto &FCallerToFAction : FCallersToFActions) {
+    Function *FCaller = FCallerToFAction.first;
+    Function *FAction = FCallerToFAction.second;
+
+    // Create a new function type with an additional function pointer argument
+    SmallVector<Type *, 4> NewParamTypes;
+    Type *FActionType = FAction->getType();
+    NewParamTypes.push_back(
+        PointerType::getUnqual(FActionType)); // Add function pointer to FAction
+    FunctionType *OldFCallerType = FCaller->getFunctionType();
+    for (Type *Ty : OldFCallerType->params())
+      NewParamTypes.push_back(Ty);
+
+    auto *NewFCallerType =
+        FunctionType::get(OldFCallerType->getReturnType(), NewParamTypes,
+                          OldFCallerType->isVarArg());
+
+    // Create a new function with the updated type and rename it to
+    // call_if_on_device_conditionally_GUID_N
+    if (!UniquePrefixOpt.empty())
+      UniquePrefix = UniquePrefixOpt;
+    Twine NewFCallerName = Twine(FCaller->getName()) + "_" + UniquePrefix +
+                           "_" + Twine(FCallerIndex);
+    // Also change to external linkage
+    Function *NewFCaller = Function::Create(
+        NewFCallerType, Function::ExternalLinkage, NewFCallerName, &M);
+
+    NewFCaller->setCallingConv(FCaller->getCallingConv());
+
+    // Replace all calls to the old function with the new one
+    for (auto &U : FCaller->uses()) {
+      if (auto *Call = dyn_cast<CallInst>(U.getUser())) {
```

Comment:
This could be turned into an early exit, thus removing a single level of indentation
```suggestion
      auto *Call = dyn_cast<CallInst>(U.getUser());
      if (!Call)
        continue;
```